icu4c/source/i18n/number_longnames.cpp

```diff
@@ -232,15 +282,96 @@ UnicodeString getPerUnitFormat(const Locale& locale, const UNumberUnitWidth &wid
     return UnicodeString(ptr, len);
 }
 
+// Load data originating from a deriveComponent CLDR element. The feature and
+// structure parameters must be null-terminated.
+void getDeriveComponentRule(Locale locale,
+                            const char *feature,
+                            const char *structure,
+                            UnicodeString &value0,
+                            UnicodeString &value1,
+                            UErrorCode &status) {
+    StackUResourceBundle derivationsBundle, stackBundle;
+    ures_openDirectFillIn(derivationsBundle.getAlias(), NULL, "grammaticalFeatures", &status);
+    ures_getByKey(derivationsBundle.getAlias(), "grammaticalData", derivationsBundle.getAlias(),
+                  &status);
+    ures_getByKey(derivationsBundle.getAlias(), "derivations", derivationsBundle.getAlias(), &status);
+    // TODO: use standard normal locale resolution algorithms rather than just grabbing language:
+    ures_getByKey(derivationsBundle.getAlias(), locale.getLanguage(), stackBundle.getAlias(), &status);
+    // TODO:
+    // - code currently assumes if the locale exists, the rules are there -
+    //   instead of falling back to root when the requested rule is missing.
+    // - investigate ures.h functions, see if one that uses res_findResource()
+    //   might be better (or use res_findResource directly), or maybe help
+    //   improve ures documentation to guide function selection?
+    if (status == U_MISSING_RESOURCE_ERROR) {
+        status = U_ZERO_ERROR;
+        ures_getByKey(derivationsBundle.getAlias(), "root", stackBundle.getAlias(), &status);
+    }
+    ures_getByKey(stackBundle.getAlias(), "component", stackBundle.getAlias(), &status);
+    ures_getByKey(stackBundle.getAlias(), feature, stackBundle.getAlias(), &status);
+    ures_getByKey(stackBundle.getAlias(), structure, stackBundle.getAlias(), &status);
+    value0 = ures_getUnicodeStringByIndex(stackBundle.getAlias(), 0, &status);
+    value1 = ures_getUnicodeStringByIndex(stackBundle.getAlias(), 1, &status);
+}
+
+UnicodeString
+getDeriveCompoundRule(Locale locale, const char *feature, const char *structure, UErrorCode &status) {
```

Comment:
As I mentioned before, I prefer to add those kind of function an a separate class.

Reply:
In Java: separate class, separate file (or nested class). In C++: this is just a function. The .cpp files tend to get rather big sometimes - but that's just file length, not really complexity?

The LongNameHandler java class has:
- PluralTableSink (subclass)
- getMeasureData
- getCurrencyLongNameData
- getPerUnitFormat

This class looks like a good place for getDeriveCompoundRule to me. Component and compound rules are all related to producing names for units.